streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java

```diff
@@ -68,72 +69,115 @@ private static synchronized void updateNumRecordsProduces(final int delta) {
     }
 
     static void generate(final String kafka) {
+        try {
+            Exit.addShutdownHook("streams-eos-test-driver-shutdown-hook", () -> {
```

Comment:
The previous shutdown hook did not wait until the "main loop" breaks and exits. Hence, the code after the loop was never executed making debugging harder. We introduce the `terminated` flag to delay the termination of the JVM until the method finished.